### ISSUE TITLE
Preprocess

### ISSUE THREAD

**Issue by mariana-yn** [UKPLab/sentence-transformers#163]
Hello,

First of all, thank you for the great paper and package, both are really great!

Sorry if those are simple questions, but I started using BERT recently.

So, I'm using a custom triplets dataset in portuguese to fine-tune a BERT model (also in portuguese, available with huggingface) and I have a doubt about the preprocessing.
Based on your script `training_wikipedia_sections.py `, when I use the `DataLoader `function, does it preprocess the texts contained on my datasets or they should already be preprocessed? Or is there another function to do that?

Also, I'm trying many variations during the training and I take almost 1 hour to load the data every time I run the script. Is there a way to save the data and load it later without using the `Dataloader`?

Just one more thing, is it possible to fine-tune BERT using the triplets dataset and get word embeddings instead of sentence embeddings? Maybe removing the pooling layer... I wanted to compare both.

Thank you!

**Comment by nreimers** [comment]
Hi @mariana-yn 
When you use the DataLoader function, it pre-processes the data, i.e., you feed in 3 sentences (or texts). The code will then handle tokenization and mapping of token & subword-tokens to the correct index in the BERT dictionary.

Sadly this takes some time, as the tokenizer is not the fastest. But HuggingFace is currently working on improving tokenization speed.

Out-of-the box there is no option to store a pre-processed dataset. But you could extend the SentencesDataset and store self.tokens and self.labels on disc. You could then load those values from disc again.

But be careful. The different models use different tokenization strategies and have different vocabs. So if you pre-process your dataset with BERT, you cannot use it with RoBERTa.

I will add an option to output word embeddings soon.

Best
Nils Reimers

**Comment by nreimers** [comment]
Hi,
I just committed an option to return the token embeddings:
```
from sentence_transformers import SentenceTransformer
model = SentenceTransformer('bert-base-nli-mean-tokens')
sentences = ['This framework generates embeddings for each input sentence',
             'Sentences are passed as a list of string.',
             'The quick brown fox jumps over the lazy dog.']

sentence_embeddings = model.encode(sentences, output_value='token_embeddings')
print(sentence_embeddings[1]) #Token embeddings for the second sentence. It includes values for the padding tokens, which are all zero
```

I hope this helps.

**Comment by mariana-yn** [comment]
That's great!
I'll try to store those values like you suggested and test the word embeddings.

Thank you very much!

**Comment by mariana-yn** [comment]
Hi @nreimers , I just have another few questions...

When I use the token_embeddings option of the model.encode, does it perform padding on the sentences? I notice that the maximum number of items is 130, só I wanted to know if it's possible to configure the padding, if it is really applied. Also, is there a way to output the wordpieces after tokenizing the text? I can use the model.tokenize to see the numeric values, but is there a way to see the text too?

Another thing that I wanted to know is if the model.encode preprocess the text the same way that the Dataloader does, since in your examples you input a not processed text and it returns the embedded version of it.

Thank you again!

**Comment by nreimers** [comment]
Hi @mariana-yn 
Yes, padding is applied on a mini-batch scale. The system checks what is the longest sentence in a mini-batch and then pads all sentences in that mini-batch to the same lenght.

There is also a cut-off at 128 word pieces, sentences longer than that will be truncated. BERT has a quadratic memory and runtime dependence on the sentence length. If sentences get too long, it will consume a lot of memory and will be quite slow.

The max sentence length can be configured in the model, for example, in the BERT model.

Out of the box no code is included to map a text to word-pieces. But it would be quite easy to get this information. You can map the IDs back to their respective entries in the vocab of the tokenizer.

Or you use this code:
```
tokenizer = BertTokenizer.from_pretrained(model_name_or_path, do_lower_case=True)
tokenizer.tokenize(text)
```

Best
Nils Reimers

**Comment by BelhalK** [comment]
Hello,

Following that discussion, I believe 
```
sentence_model = SentenceTransformer('paraphrase-multilingual-mpnet-base-v2')
tokens = sentence_model.tokenize(mybatchofsentences)
```
will do padding using the max length in the minibtach.
Yet, I am wondering if 
```
emb = sentence_model(tokens)['sentence_embedding']
```
returns the embeddings vector after doing a mean over all embeddings (even those computed from 0 padded tokens) or only the non zero token embeddings?

Thank you for your help.
Belhal
